python/ray/experimental/sgd.md

```diff
@@ -0,0 +1,401 @@
+SGD API Review
+==============
+
+There are a couple use cases and features we want to support for SGD.
+
+1. As a user, I want to speed up my single-threaded model training code by distributed training.
+2. As a user, I want to be able to feed in a large dataset (1TB) from the cloud into my distributed training scheme.
+3. As a user, I want to be able to tune hyperparameters for any distributed training set up.
+4. As a user, I want to be able to do elastic distributed training, (including with a large dataset [P1]).
+5. As a user, I want to be able to develop/train my model on a jupyter notebook and validate it in the same environment.
+6. As a user, I want to be able to integrate RaySGD in a way that is easily compatible with single-threaded code.
+7. As a user, I want to be able to monitor the performance of my training mode.
+8. As a user, I want my training to be completely reproducible.
+6. [P1] As a user, I want to be able to have a plug-and-play solution for optimally speeding up my model training code / supporting large models.
+
+
+### Non-supported use cases for Alpha 
+
+1. Elastic training + large dataset (not yet, beta)
+2. Supporting multiple datasets.
+
+Below, I map out the APIs that meet each of the above requirements (R1, R2, R3, etc). 
+
+General Model Training 
+----------------------
+
+Satisfies: R1, R5
+Maybe satisfies: R6. 
+
+Previous RaySGD API forced you to refactor your code. This is very high burden for users.
+
+**One single long-running execution**
+```python
+
+def train_func(config):  # user always needs "config"
+    model = Net()
+    optimizer = torch.optimizer.SGD()
+    model.cuda()
+    DistributedDataParallel(...)
+
+    # ray.sgd._local_rank() # not sure if we should expose
```

Comment:
We still want to support cases where users might have specific logic based on rank. For example, they want to download the data, but only with 1 worker per node. Or they want to setup some logging, but only the rank 0 worker should do this.

Reply:
+1; maybe we can have a `is_master` and/or `is_local_master`.